src/cargo/core/workspace.rs

```diff
@@ -506,13 +528,43 @@ impl<'gctx> Workspace<'gctx> {
     /// Returns the root `[patch]` section of this workspace.
     ///
     /// This may be from a virtual crate or an actual crate.
-    pub fn root_patch(&self) -> CargoResult<HashMap<Url, Vec<Dependency>>> {
-        let from_manifest = match self.root_maybe() {
+    pub fn root_patch(&self) -> CargoResult<HashMap<Url, Vec<(Dependency, PathOrDefinition)>>> {
+        let patch_cv = self.gctx.get_cv_with_env(&ConfigKey::from_str("patch"))?;
+        let manifest_path = PathOrDefinition::Path(self.root_manifest().into());
+
+        let patch_from_manifest = match self.root_maybe() {
             MaybePackage::Package(p) => p.manifest().patch(),
             MaybePackage::Virtual(vm) => vm.patch(),
         };
 
-        let from_config = self.config_patch()?;
+        // Insert location information for each dep in the patch
+        let mut from_config: HashMap<_, Vec<_>> = HashMap::new();
+        for (url, dep) in self.config_patch()? {
+            from_config.insert(
+                url,
+                dep.iter()
+                    .map(|dep| {
+                        (
+                            dep.clone(),
+                            PathOrDefinition::Definition(
+                                patch_cv.as_ref().unwrap().definition().clone(),
+                            ),
+                        )
+                    })
+                    .collect(),
+            );
+        }
+
+        let mut from_manifest: HashMap<_, Vec<_>> = HashMap::new();
+        for (url, dep) in patch_from_manifest {
+            from_manifest.insert(
+                url.clone(),
+                dep.iter()
+                    .map(|dep| (dep.clone(), manifest_path.clone()))
+                    .collect(),
+            );
+        }
```

Comment:
Should this be done in the functions we call to get a patch?  Please understand, this is a serious question and not an indirect way of saying what to do.

For manifests, its a little weird because it will always be the same path.  However, it feels weird to be calling back into config to get the location.  It feels like we are re-implementing things.  Calling back into config can also have some negative performance implications.

Reply:
Not sure if this is what you said, but maybe it's better to move the` get_cv_with_env `call to  `Workspace::config_patch` as it's already calling the config there with the following: 
```
        let config_patch: Option<
            BTreeMap<String, BTreeMap<String, TomlDependency<ConfigRelativePath>>>,
        > = self.gctx.get("patch")?;
```
But it would still have the negative performance implications. I'm not sure if there's a better way to get the config location.